fqdn: Fix notifyOnDNSMsg benchmark

pippolo84 opened the pull request:
The design of BenchmarkNotifyOnDNSMsg was flawed due to the usage of b.N as benchmark input size. b.N is used by the Go benchmarking framework to find a reliable timing for the code under test. Changing the amount of work done at each benchmark call leads to unreliable numbers.  For more info refer to the Go documentation:

	The benchmark function must run the target code b.N times. During
	benchmark execution, b.N is adjusted until the benchmark function
	lasts long enough to be timed reliably.

And the section `Traps for young players` in the blog post https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go

To fix it, the benchmark has been updated to setup a reasonable fixed number of endpoints.

Also, since the code was behind the latest development in the policy and fqdn subsystems, it has been updated to run correctly. Specifically:

- the global LocalNodeStore is set once at startup
- the FQDN selectors are added to the selector cache with a dummy user
- the daemon context is properly initialized
- the FQDN related global opts are set to their defaults to avoid noisy logs

Example run:

```
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/daemon/cmd
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkNotifyOnDNSMsg-8   	      21	  47992143 ns/op	30757230 B/op	  357646 allocs/op
BenchmarkNotifyOnDNSMsg-8   	      24	  48773334 ns/op	30590877 B/op	  355639 allocs/op
BenchmarkNotifyOnDNSMsg-8   	      22	  47940489 ns/op	30719252 B/op	  358088 allocs/op
BenchmarkNotifyOnDNSMsg-8   	      24	  48019454 ns/op	30364074 B/op	  353470 allocs/op
BenchmarkNotifyOnDNSMsg-8   	      24	  66607283 ns/op	30634838 B/op	  356352 allocs/op
PASS
```

Related: https://github.com/cilium/cilium/pull/32297#discussion_r1588937381

pippolo84 committed:
fqdn: Fix notifyOnDNSMsg benchmark

The design of BenchmarkNotifyOnDNSMsg was flawed due to the usage of b.N
as benchmark input size. b.N is used by the Go benchmarking framework to
find a reliable timing for the code under test. Changing the amount of
work done at each benchmark call leads to unreliable numbers.  For more
info refer to the Go documentation:

	The benchmark function must run the target code b.N times. During
	benchmark execution, b.N is adjusted until the benchmark function
	lasts long enough to be timed reliably.

And the section `Traps for young players` in the blog post
https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go

To fix it, the benchmark has been updated to setup a reasonable fixed
number of endpoints.

Also, since the code was behind the latest development in the policy and
fqdn subsystems, it has been updated to run correctly. Specifically:

- the global LocalNodeStore is set once at startup
- the FQDN selectors are added to the selector cache with a dummy user
- the daemon context is properly initialized
- the FQDN related global opts are set to their defaults to avoid noisy
  logs

Example run:

goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/daemon/cmd
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkNotifyOnDNSMsg-8   	      21	  47992143 ns/op	30757230 B/op	  357646 allocs/op
BenchmarkNotifyOnDNSMsg-8   	      24	  48773334 ns/op	30590877 B/op	  355639 allocs/op
BenchmarkNotifyOnDNSMsg-8   	      22	  47940489 ns/op	30719252 B/op	  358088 allocs/op
BenchmarkNotifyOnDNSMsg-8   	      24	  48019454 ns/op	30364074 B/op	  353470 allocs/op
BenchmarkNotifyOnDNSMsg-8   	      24	  66607283 ns/op	30634838 B/op	  356352 allocs/op
PASS

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

doniacld commented on daemon/cmd/fqdn_test.go:
```diff
@@ -37,13 +39,32 @@ type DaemonFQDNSuite struct {
+	notifyOnDNSMsgBenchSetup.Do(func() {
+		// set FQDN related options to defaults in order to avoid a flood of warnings
+		option.Config.DNSProxyLockCount = 131
+		option.Config.DNSProxyLockTimeout = 500 * time.Millisecond
```
nit: Don't we have constants for these values?

pippolo84 replied:
Added them in a separate commit :+1:

doniacld reviewed (approved):
One nit comment, otherwise lgtm 🚀 
NB: Particularly appreciate the commit message.

pippolo84 committed:
fqdn: Add default values for DNSProxyLock{Count,Timeout}

Add defaults values for DNSProxyLockCount and DNSProxyLockTimeout in
order to refer them both in the global options initialization and in the
notifyOnDNSMessage benchmark.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

vipul-21 commented on daemon/cmd/fqdn_test.go:
```diff
@@ -92,21 +114,22 @@ func BenchmarkNotifyOnDNSMsg(b *testing.B) {
 	selectorsToAdd := api.FQDNSelectorSlice{ciliumIOSel, ciliumIOSelMatchPattern, ebpfIOSel}
-	nameManager.Lock()
 	for _, sel := range selectorsToAdd {
-		nameManager.RegisterForIPUpdatesLocked(sel)
```
Just curious, why are we not adding the fqdnselector via `namemanager` and directly though selector cache ?

pippolo84 replied:
We need to do that via the SelectorCache to create an `identitySelector` struct and initialize its `source` field:

https://github.com/cilium/cilium/blob/5f6456a4d2190752d1e3955978f55d5a21a75b98/pkg/policy/selectorcache.go#L404-L411

that will be referenced in:

https://github.com/cilium/cilium/blob/5f6456a4d2190752d1e3955978f55d5a21a75b98/pkg/policy/selectorcache.go#L310

when we update the FQDN selector from `notifyOnDNSMsg`.

doniacld reviewed (approved):
Good! Thanks for the changes!
